### ISSUE TITLE
Feedback for CodeGen

### ISSUE THREAD

**Issue by magnus-madsen** [flix/flix#153]
I merged the CodeGen pull request, but I had some minor comments:
(The order of comments is insubstantial.)
- [x] The codegen backend is meant to be THE BACKEND. So let us go ahead and (1) make it the default. (2) remove the "-c" and "--compile" flags! 
- [x] On that subject, let us adopt Java's convention on command line arguments. We have for example `-d`, `--debugger`, `--verify` which is good. On the other hand, we should hide flags that are not public and mostly for internal use. For example, the code generator will be default, so to run in interpreted mode, you will supply the argument `-Xinterpreter`. Similarly, you can have `-Xdump-bytecode`. 
- [ ] We should make sure everything after TypedAst matches the structure of "WeededAst" especially w.r.t to the names "Declaration", "Definition", etc. Can you try to bring it in sync?
- [ ] A minor thing, I prefer to use `:::` when concatenating lists instead of `++` which works for any type. This can sometimes avoid minor bugs when mixing types. (e.g. lambda lift, line 57)
- [ ] LambdaLift, freeVars, why use a mutable map? 
- [x] What does this mean `import org.objectweb.asm.{Type => _, _}`?
- [ ] Let us refactor Value soon, but let's do it in a separate PR and branch. I think that would clean up the codegen quiet a bit! :)
- [x] I forget, why is it called `MkClosureRef` and not just `MkClosure`?
- [ ] CodeGen line 213. I see creating an Expression.Var allows some nice code-reuse. On creating AST fragments can be problematic. Especially because soon I will introduce a unique identifier for every AST node, and so that will make things a bit more complicated. Let us talk about this one. For now its okay.
  -[ ] For equality, we need to come up with some general mechanism. Possibly there will be some interface or static class.

**Comment by magnus-madsen** [comment]
What is the 'status' on this?

**Comment by mhyee** [comment]
I edited the original post to use checkboxes so we can see what's actually left to do. (Note that some of those comments were questions, which I answered previously.)

Remaining things for you to answer/respond (some of my response copied from above, and those responses need further responses from you):

> We should make sure everything after TypedAst matches the structure of "WeededAst" especially w.r.t to the names "Declaration", "Definition", etc. Can you try to bring it in sync?

On taking a closer look, this is a very large task, with some in-progress work (e.g. removing Lit and Term). We should probably discuss before making large changes here.

We should also open a separate issue.

> A minor thing, I prefer to use ::: when concatenating lists instead of ++ which works for any type. This can sometimes avoid minor bugs when mixing types. (e.g. lambda lift, line 57)

I think the line numbers have changed, so where specifically is this? Or is this something general? 

> LambdaLift, freeVars, why use a mutable map?

Are you referring to method freeVariables in ClosureConv? There is no immutable LinkedHashSet.

Or is this the mutable map of definitions (type TopLevel) in LambdaLift? It's easier for the lift method to mutate a parameter, instead of returning a pair of (Expression, TopLevel).

Or something else? There's freeVars, used as a variable and a field of an AST node, but it's an immutable list.

> Let us refactor Value soon, but let's do it in a separate PR and branch. I think that would clean up the codegen quiet a bit! :)

This is making Value more Java-like, right? We should open a separate issue. But I'm not sure this will be as helpful as we had believed. We can't completely eliminate strings in the code generator, but we've gotten pretty close. We could probably do better if we made Value more Java-like, but that might not be worth it. (Too much work, and then Value is less organized / maintainable.)

> CodeGen line 213. I see creating an Expression.Var allows some nice code-reuse. On creating AST fragments can be problematic. Especially because soon I will introduce a unique identifier for every AST node, and so that will make things a bit more complicated. Let us talk about this one. For now its okay.

There are similar cases in the Interpreter and Codegen, when I rewrite e1 ==> e2 into !e1 || e2 and e1 <==> e2 into e1 == e2.

But no real resolution right now. Creating AST nodes makes it easier and less redundant to implement certain expressions.

> For equality, we need to come up with some general mechanism. Possibly there will be some interface or static class.

There's java.lang.Object.equals(), but the implementation is messier because of the different primitive types and optimizations (e.g. reference equality for interned types, directly calling java.util.Arrays.equals for tuples).

See also comment [here](https://github.com/flix/flix/issues/165#issuecomment-230660081). We don't really have a good place to discuss this, so we should probably open a separate issue for equality (since it's more general than Value.Tuple.)
